File: src/family-entry-contracts.ts
export type JsonRecord = Record<string, unknown>;

export type DomainEntryCommandContract = JsonRecord & {
  command: string;
  required_fields: string[];
  optional_fields: string[];
};

export type FamilyDomainEntryContractSurface = JsonRecord & {
  entry_adapter: string;
  service_safe_surface_kind: string;
  product_entry_builder_command: string;
  supported_commands: string[];
  command_contracts: DomainEntryCommandContract[];
  supported_entry_modes?: string[];
  product_entry_kind?: string;
};

export type GatewayInteractionContractSurface = JsonRecord & {
  surface_kind: string;
  frontdoor_owner: string;
  user_interaction_mode: string;
  user_commands_required: boolean;
  command_surfaces_for_agent_consumption_only: boolean;
  shared_downstream_entry: string;
  shared_handoff_envelope: string[];
};

export type SharedHandoffBuilderSurface = JsonRecord & {
  command: string;
  entry_mode: string;
  surface_kind?: string;
};

export type SharedHandoffReturnSurface = JsonRecord & {
  surface_kind: string;
  target_domain_id: string;
};

export type FamilySharedHandoffSurface = JsonRecord & {
  direct_entry_builder?: SharedHandoffBuilderSurface;
  opl_handoff_builder?: SharedHandoffBuilderSurface;
  opl_return_surface?: SharedHandoffReturnSurface;
};

export interface BuildDomainEntryCommandContractInput {
  command: string;
  required_fields: string[];
  optional_fields?: string[] | null;
  extra_payload?: JsonRecord;
}

export interface DomainEntryCommandCatalogSurface {
  supported_commands: string[];
  command_contracts: DomainEntryCommandContract[];
}

export interface BuildFamilyDomainEntryContractInput {
  entry_adapter: string;
  service_safe_surface_kind: string;
  product_entry_builder_command: string;
  supported_commands: string[];
  command_contracts: Array<DomainEntryCommandContract | JsonRecord>;
  supported_entry_modes?: string[] | null;
  product_entry_kind?: string | null;
  extra_payload?: JsonRecord;
}

export interface BuildGatewayInteractionContractInput {
  frontdoor_owner: string;
  user_interaction_mode: string;
  user_commands_required: boolean;
  command_surfaces_for_agent_consumption_only: boolean;
  shared_downstream_entry: string;
  shared_handoff_envelope: string[];
  surface_kind?: string | null;
  extra_payload?: JsonRecord;
}

export interface BuildFamilyGatewayInteractionContractInput {
  shared_downstream_entry: string;
  extra_shared_handoff_envelope?: string[] | null;
  frontdoor_owner?: string | null;
  user_interaction_mode?: string | null;
  user_commands_required?: boolean | null;
  command_surfaces_for_agent_consumption_only?: boolean | null;
  surface_kind?: string | null;
  extra_payload?: JsonRecord;
}

export interface BuildSharedHandoffBuilderInput {
  command: string;
  entry_mode: string;
  surface_kind?: string | null;
  extra_payload?: JsonRecord;
}

export interface BuildSharedHandoffReturnSurfaceInput {
  surface_kind: string;
  target_domain_id: string;
  extra_payload?: JsonRecord;
}

export interface BuildSharedHandoffInput {
  direct_entry_builder?: SharedHandoffBuilderSurface | JsonRecord | null;
  opl_handoff_builder?: SharedHandoffBuilderSurface | JsonRecord | null;
  opl_return_surface?: SharedHandoffReturnSurface | JsonRecord | null;
  extra_payload?: JsonRecord;
}

export interface BuildFamilyDirectOplSharedHandoffInput {
  direct_entry_builder_command: string;
  opl_handoff_builder_command: string;
  direct_entry_mode?: string | null;
  opl_handoff_entry_mode?: string | null;
  extra_payload?: JsonRecord;
}

const SHARED_HANDOFF_KEYS = [
  'direct_entry_builder',
  'opl_handoff_builder',
  'opl_return_surface',
] as const;

export const DEFAULT_FAMILY_GATEWAY_SHARED_HANDOFF_ENVELOPE = [
  'target_domain_id',
  'task_intent',
  'entry_mode',
  'workspace_locator',
  'runtime_session_contract',
  'return_surface_contract',
] as const;

function isRecord(value: unknown): value is JsonRecord {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function optionalString(value: unknown) {
  return typeof value === 'string' && value.trim() ? value.trim() : null;
}

function requireString(value: unknown, field: string) {
  const text = optionalString(value);
  if (!text) {
    throw new Error(`family entry contract 缺少字符串字段: ${field}`);
  }
  return text;
}

function requireBoolean(value: unknown, field: string) {
  if (typeof value !== 'boolean') {
    throw new Error(`family entry contract 缺少布尔字段: ${field}`);
  }
  return value;
}

function requireRecord(value: unknown, field: string) {
  if (!isRecord(value)) {
    throw new Error(`family entry contract 缺少对象字段: ${field}`);
  }
  return value;
}

function readStringList(value: unknown, field: string) {
  if (!Array.isArray(value)) {
    throw new Error(`family entry contract 缺少数组字段: ${field}`);
  }
  return value.map((entry, index) => requireString(entry, `${field}[${index}]`));
}

function readOptionalStringList(value: unknown, field: string) {
  if (value === undefined || value === null) {
    return undefined;
  }
  return readStringList(value, field);
}

function mergeExtraPayload(base: JsonRecord, extraPayload: unknown, field: string) {
  if (extraPayload === undefined || extraPayload === null) {
    return base;
  }
  const payload = requireRecord(extraPayload, field);
  for (const key of Object.keys(payload)) {
    if (key in base) {
      throw new Error(`family entry contract extra_payload 不允许覆盖核心字段: ${field}.${key}`);
    }
  }
  return {
    ...base,
    ...payload,
  };
}

function omitKeys(payload: JsonRecord, keys: string[]) {
  const omitted = new Set(keys);
  return Object.fromEntries(Object.entries(payload).filter(([key]) => !omitted.has(key)));
}

export function buildDomainEntryCommandContract(
  input: BuildDomainEntryCommandContractInput,
): DomainEntryCommandContract {
  return mergeExtraPayload(
    {
      command: requireString(input.command, 'command'),
      required_fields: readStringList(input.required_fields, 'required_fields'),
      optional_fields: readOptionalStringList(input.optional_fields, 'optional_fields') ?? [],
    },
    input.extra_payload,
    'command_contract',
  ) as DomainEntryCommandContract;
}

export function buildDomainEntryCommandCatalog(
  inputs: BuildDomainEntryCommandContractInput[],
): DomainEntryCommandCatalogSurface {
  const commandContracts = inputs.map((input, index) =>
    buildDomainEntryCommandContract({
      command: requireString(input.command, `command_catalog[${index}].command`),
      required_fields: readStringList(
        input.required_fields,
        `command_catalog[${index}].required_fields`,
      ),
      optional_fields:
        readOptionalStringList(
          input.optional_fields,
          `command_catalog[${index}].optional_fields`,
        ) ?? [],
      extra_payload: input.extra_payload,
    }),
  );

  return {
    supported_commands: commandContracts.map((contract) => contract.command),
    command_contracts: commandContracts,
  };
}

export function validateFamilyDomainEntryContract(
  value: unknown,
  field: string,
): FamilyDomainEntryContractSurface {
  const payload = requireRecord(value, field);
  const supportedCommands = readStringList(payload.supported_commands, `${field}.supported_commands`);
  const rawCommandContracts = payload.command_contracts;
  if (!Array.isArray(rawCommandContracts) || rawCommandContracts.length === 0) {
    throw new Error(`family entry contract 缺少数组字段: ${field}.command_contracts`);
  }
  const commandContracts = rawCommandContracts.map((contract, index) => {
    const normalized = requireRecord(contract, `${field}.command_contracts[${index}]`);
    const command = requireString(normalized.command, `${field}.command_contracts[${index}].command`);
    const requiredFields = readStringList(
      normalized.required_fields,
      `${field}.command_contracts[${index}].required_fields`,
    );
    const optionalFields = readOptionalStringList(
      normalized.optional_fields,
      `${field}.command_contracts[${index}].optional_fields`,
    );
    return buildDomainEntryCommandContract({
      command,
      required_fields: requiredFields,
      optional_fields: optionalFields,
      extra_payload: omitKeys(normalized, ['command', 'required_fields', 'optional_fields']),
    });
  });

  const normalized: FamilyDomainEntryContractSurface = {
    ...payload,
    entry_adapter: requireString(payload.entry_adapter, `${field}.entry_adapter`),
    service_safe_surface_kind: requireString(
      payload.service_safe_surface_kind,
      `${field}.service_safe_surface_kind`,
    ),
    product_entry_builder_command: requireString(
      payload.product_entry_builder_command,
      `${field}.product_entry_builder_command`,
    ),
    supported_commands: supportedCommands,
    command_contracts: commandContracts,
  };
  const supportedEntryModes = readOptionalStringList(
    payload.supported_entry_modes,
    `${field}.supported_entry_modes`,
  );
  if (supportedEntryModes !== undefined) {
    normalized.supported_entry_modes = supportedEntryModes;
  }
  const productEntryKind = optionalString(payload.product_entry_kind);
  if (productEntryKind) {
    normalized.product_entry_kind = productEntryKind;
  }
  return normalized;
}

export function buildFamilyDomainEntryContract(
  input: BuildFamilyDomainEntryContractInput,
): FamilyDomainEntryContractSurface {
  const base: JsonRecord = {
    entry_adapter: requireString(input.entry_adapter, 'entry_adapter'),
    service_safe_surface_kind: requireString(
      input.service_safe_surface_kind,
      'service_safe_surface_kind',
    ),
    product_entry_builder_command: requireString(
      input.product_entry_builder_command,
      'product_entry_builder_command',
    ),
    supported_commands: readStringList(input.supported_commands, 'supported_commands'),
    command_contracts: input.command_contracts.map((contract, index) => {
      const normalized = requireRecord(contract, `command_contracts[${index}]`);
      const command = requireString(normalized.command, `command_contracts[${index}].command`);
      const requiredFields = readStringList(
        normalized.required_fields,
        `command_contracts[${index}].required_fields`,
      );
      const optionalFields = readOptionalStringList(
        normalized.optional_fields,
        `command_contracts[${index}].optional_fields`,
      );
      return buildDomainEntryCommandContract({
        command,
        required_fields: requiredFields,
        optional_fields: optionalFields,
        extra_payload: omitKeys(
          normalized,
          ['command', 'required_fields', 'optional_fields'],
        ),
      });
    }),
  };
  const supportedEntryModes = readOptionalStringList(input.supported_entry_modes, 'supported_entry_modes');
  if (supportedEntryModes !== undefined) {
    base.supported_entry_modes = supportedEntryModes;
  }
  const productEntryKind = optionalString(input.product_entry_kind);
  if (productEntryKind) {
    base.product_entry_kind = productEntryKind;
  }
  return validateFamilyDomainEntryContract(
    mergeExtraPayload(base, input.extra_payload, 'domain_entry_contract'),
    'domain_entry_contract',
  );
}

export function validateGatewayInteractionContract(
  value: unknown,
  field: string,
): GatewayInteractionContractSurface {
  const payload = requireRecord(value, field);
  return {
    ...payload,
    surface_kind: requireString(payload.surface_kind, `${field}.surface_kind`),
    frontdoor_owner: requireString(payload.frontdoor_owner, `${field}.frontdoor_owner`),
    user_interaction_mode: requireString(
      payload.user_interaction_mode,
      `${field}.user_interaction_mode`,
    ),
    user_commands_required: requireBoolean(
      payload.user_commands_required,
      `${field}.user_commands_required`,
    ),
    command_surfaces_for_agent_consumption_only: requireBoolean(
      payload.command_surfaces_for_agent_consumption_only,
      `${field}.command_surfaces_for_agent_consumption_only`,
    ),
    shared_downstream_entry: requireString(
      payload.shared_downstream_entry,
      `${field}.shared_downstream_entry`,
    ),
    shared_handoff_envelope: readStringList(
      payload.shared_handoff_envelope,
      `${field}.shared_handoff_envelope`,
    ),
  };
}

export function buildGatewayInteractionContract(
  input: BuildGatewayInteractionContractInput,
): GatewayInteractionContractSurface {
  return validateGatewayInteractionContract(
    mergeExtraPayload(
      {
        surface_kind: optionalString(input.surface_kind) ?? 'gateway_interaction_contract',
        frontdoor_owner: requireString(input.frontdoor_owner, 'frontdoor_owner'),
        user_interaction_mode: requireString(input.user_interaction_mode, 'user_interaction_mode'),
        user_commands_required: requireBoolean(
          input.user_commands_required,
          'user_commands_required',
        ),
        command_surfaces_for_agent_consumption_only: requireBoolean(
          input.command_surfaces_for_agent_consumption_only,
          'command_surfaces_for_agent_consumption_only',
        ),
        shared_downstream_entry: requireString(
          input.shared_downstream_entry,
          'shared_downstream_entry',
        ),
        shared_handoff_envelope: readStringList(
          input.shared_handoff_envelope,
          'shared_handoff_envelope',
        ),
      },
      input.extra_payload,
      'gateway_interaction_contract',
    ),
    'gateway_interaction_contract',
  );
}

export function buildFamilyGatewayInteractionContract(
  input: BuildFamilyGatewayInteractionContractInput,
): GatewayInteractionContractSurface {
  const extraEnvelope = readOptionalStringList(
    input.extra_shared_handoff_envelope,
    'extra_shared_handoff_envelope',
  ) ?? [];
  const sharedHandoffEnvelope = Array.from(
    new Set([
      ...DEFAULT_FAMILY_GATEWAY_SHARED_HANDOFF_ENVELOPE,
      ...extraEnvelope,
    ]),
  );
  return buildGatewayInteractionContract({
    frontdoor_owner: optionalString(input.frontdoor_owner) ?? 'opl_gateway_or_domain_gui',
    user_interaction_mode: optionalString(input.user_interaction_mode) ?? 'natural_language_frontdoor',
    user_commands_required: input.user_commands_required ?? false,
    command_surfaces_for_agent_consumption_only:
      input.command_surfaces_for_agent_consumption_only ?? true,
    shared_downstream_entry: requireString(
      input.shared_downstream_entry,
      'shared_downstream_entry',
    ),
    shared_handoff_envelope: sharedHandoffEnvelope,
    surface_kind: optionalString(input.surface_kind),
    extra_payload: input.extra_payload,
  });
}

export function validateSharedHandoffBuilder(
  value: unknown,
  field: string,
): SharedHandoffBuilderSurface {
  const payload = requireRecord(value, field);
  const normalized: SharedHandoffBuilderSurface = {
    ...payload,
    command: requireString(payload.command, `${field}.command`),
    entry_mode: requireString(payload.entry_mode, `${field}.entry_mode`),
  };
  const surfaceKind = optionalString(payload.surface_kind);
  if (surfaceKind) {
    normalized.surface_kind = surfaceKind;
  }
  return normalized;
}

export function buildSharedHandoffBuilder(
  input: BuildSharedHandoffBuilderInput,
): SharedHandoffBuilderSurface {
  return validateSharedHandoffBuilder(
    mergeExtraPayload(
      {
        command: requireString(input.command, 'command'),
        entry_mode: requireString(input.entry_mode, 'entry_mode'),
        ...(optionalString(input.surface_kind)
          ? { surface_kind: optionalString(input.surface_kind) }
          : {}),
      },
      input.extra_payload,
      'shared_handoff_builder',
    ),
    'shared_handoff_builder',
  );
}

export function validateSharedHandoffReturnSurface(
  value: unknown,
  field: string,
): SharedHandoffReturnSurface {
  const payload = requireRecord(value, field);
  return {
    ...payload,
    surface_kind: requireString(payload.surface_kind, `${field}.surface_kind`),
    target_domain_id: requireString(payload.target_domain_id, `${field}.target_domain_id`),
  };
}

export function buildSharedHandoffReturnSurface(
  input: BuildSharedHandoffReturnSurfaceInput,
): SharedHandoffReturnSurface {
  return validateSharedHandoffReturnSurface(
    mergeExtraPayload(
      {
        surface_kind: requireString(input.surface_kind, 'surface_kind'),
        target_domain_id: requireString(input.target_domain_id, 'target_domain_id'),
      },
      input.extra_payload,
      'shared_handoff_return_surface',
    ),
    'shared_handoff_return_surface',
  );
}

export function validateSharedHandoff(
  value: unknown,
  field: string,
): FamilySharedHandoffSurface {
  const payload = requireRecord(value, field);
  const normalized: FamilySharedHandoffSurface = {
    ...payload,
  };
  let hasKnownSurface = false;

  if (payload.direct_entry_builder !== undefined) {
    normalized.direct_entry_builder = validateSharedHandoffBuilder(
      payload.direct_entry_builder,
      `${field}.direct_entry_builder`,
    );
    hasKnownSurface = true;
  }
  if (payload.opl_handoff_builder !== undefined) {
    normalized.opl_handoff_builder = validateSharedHandoffBuilder(
      payload.opl_handoff_builder,
      `${field}.opl_handoff_builder`,
    );
    hasKnownSurface = true;
  }
  if (payload.opl_return_surface !== undefined) {
    normalized.opl_return_surface = validateSharedHandoffReturnSurface(
      payload.opl_return_surface,
      `${field}.opl_return_surface`,
    );
    hasKnownSurface = true;
  }

  if (!hasKnownSurface) {
    throw new Error(`family entry contract shared_handoff 至少需要一个已知 surface: ${field}`);
  }

  return normalized;
}

export function buildSharedHandoff(
  input: BuildSharedHandoffInput,
): FamilySharedHandoffSurface {
  const base: JsonRecord = {};

  for (const key of SHARED_HANDOFF_KEYS) {
    const value = input[key];
    if (value === undefined || value === null) {
      continue;
    }
    if (key === 'opl_return_surface') {
      base[key] = validateSharedHandoffReturnSurface(
        value,
        `shared_handoff.${key}`,
      );
      continue;
    }
    base[key] = validateSharedHandoffBuilder(value, `shared_handoff.${key}`);
  }

  return validateSharedHandoff(
    mergeExtraPayload(base, input.extra_payload, 'shared_handoff'),
    'shared_handoff',
  );
}

export function buildFamilyDirectOplSharedHandoff(
  input: BuildFamilyDirectOplSharedHandoffInput,
): FamilySharedHandoffSurface {
  return buildSharedHandoff({
    direct_entry_builder: buildSharedHandoffBuilder({
      command: requireString(
        input.direct_entry_builder_command,
        'direct_entry_builder_command',
      ),
      entry_mode: optionalString(input.direct_entry_mode) ?? 'direct',
    }),
    opl_handoff_builder: buildSharedHandoffBuilder({
      command: requireString(
        input.opl_handoff_builder_command,
        'opl_handoff_builder_command',
      ),
      entry_mode: optionalString(input.opl_handoff_entry_mode) ?? 'opl-handoff',
    }),
    extra_payload: input.extra_payload,
  });
}

File: python/opl-harness-shared/src/opl_harness_shared/family_entry_contracts.py
from __future__ import annotations

from typing import Any, Mapping


def _non_empty_text(value: object) -> str | None:
    text = str(value or "").strip()
    return text or None


def _require_string(value: object, field: str) -> str:
    text = _non_empty_text(value)
    if text is None:
        raise ValueError(f"family entry contract 缺少字符串字段: {field}")
    return text


def _require_bool(value: object, field: str) -> bool:
    if not isinstance(value, bool):
        raise ValueError(f"family entry contract 缺少布尔字段: {field}")
    return value


def _require_mapping(value: object, field: str) -> Mapping[str, Any]:
    if not isinstance(value, Mapping):
        raise ValueError(f"family entry contract 缺少对象字段: {field}")
    return value


def _require_string_list(value: object, field: str) -> list[str]:
    if not isinstance(value, list):
        raise ValueError(f"family entry contract 缺少数组字段: {field}")
    return [_require_string(entry, f"{field}[{index}]") for index, entry in enumerate(value)]


def _read_optional_string_list(value: object, field: str) -> list[str] | None:
    if value is None:
        return None
    return _require_string_list(value, field)


def _merge_extra_payload(base: dict[str, Any], extra_payload: object | None, *, field: str) -> dict[str, Any]:
    if extra_payload is None:
        return base
    payload = dict(_require_mapping(extra_payload, field))
    for key in payload:
        if key in base:
            raise ValueError(f"family entry contract extra_payload 不允许覆盖核心字段: {field}.{key}")
    return {
        **base,
        **payload,
    }


def _omit_keys(payload: Mapping[str, Any], keys: tuple[str, ...]) -> dict[str, Any]:
    excluded = set(keys)
    return {key: value for key, value in payload.items() if key not in excluded}


def build_domain_entry_command_contract(
    *,
    command: str,
    required_fields: list[str],
    optional_fields: list[str] | None = None,
    extra_payload: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    return _merge_extra_payload(
        {
            "command": _require_string(command, "command"),
            "required_fields": _require_string_list(required_fields, "required_fields"),
            "optional_fields": _read_optional_string_list(optional_fields, "optional_fields") or [],
        },
        extra_payload,
        field="command_contract",
    )


def validate_family_domain_entry_contract(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    supported_commands = _require_string_list(
        payload.get("supported_commands"),
        f"{field}.supported_commands",
    )
    raw_command_contracts = payload.get("command_contracts")
    if not isinstance(raw_command_contracts, list) or not raw_command_contracts:
        raise ValueError(f"family entry contract 缺少数组字段: {field}.command_contracts")
    command_contracts = [
        build_domain_entry_command_contract(
            command=_require_mapping(contract, f"{field}.command_contracts[{index}]").get("command"),  # type: ignore[arg-type]
            required_fields=_require_mapping(contract, f"{field}.command_contracts[{index}]").get("required_fields"),  # type: ignore[arg-type]
            optional_fields=_require_mapping(contract, f"{field}.command_contracts[{index}]").get("optional_fields"),  # type: ignore[arg-type]
            extra_payload=_omit_keys(
                _require_mapping(contract, f"{field}.command_contracts[{index}]"),
                ("command", "required_fields", "optional_fields"),
            ),
        )
        for index, contract in enumerate(raw_command_contracts)
    ]
    normalized = {
        **dict(payload),
        "entry_adapter": _require_string(payload.get("entry_adapter"), f"{field}.entry_adapter"),
        "service_safe_surface_kind": _require_string(
            payload.get("service_safe_surface_kind"),
            f"{field}.service_safe_surface_kind",
        ),
        "product_entry_builder_command": _require_string(
            payload.get("product_entry_builder_command"),
            f"{field}.product_entry_builder_command",
        ),
        "supported_commands": supported_commands,
        "command_contracts": command_contracts,
    }
    supported_entry_modes = _read_optional_string_list(
        payload.get("supported_entry_modes"),
        f"{field}.supported_entry_modes",
    )
    if supported_entry_modes is not None:
        normalized["supported_entry_modes"] = supported_entry_modes
    product_entry_kind = _non_empty_text(payload.get("product_entry_kind"))
    if product_entry_kind is not None:
        normalized["product_entry_kind"] = product_entry_kind
    return normalized


def build_family_domain_entry_contract(
    *,
    entry_adapter: str,
    service_safe_surface_kind: str,
    product_entry_builder_command: str,
    supported_commands: list[str],
    command_contracts: list[Mapping[str, Any]],
    supported_entry_modes: list[str] | None = None,
    product_entry_kind: str | None = None,
    extra_payload: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    base: dict[str, Any] = {
        "entry_adapter": _require_string(entry_adapter, "entry_adapter"),
        "service_safe_surface_kind": _require_string(
            service_safe_surface_kind,
            "service_safe_surface_kind",
        ),
        "product_entry_builder_command": _require_string(
            product_entry_builder_command,
            "product_entry_builder_command",
        ),
        "supported_commands": _require_string_list(supported_commands, "supported_commands"),
        "command_contracts": [
            build_domain_entry_command_contract(
                command=_require_mapping(contract, f"command_contracts[{index}]").get("command"),  # type: ignore[arg-type]
                required_fields=_require_mapping(contract, f"command_contracts[{index}]").get("required_fields"),  # type: ignore[arg-type]
                optional_fields=_require_mapping(contract, f"command_contracts[{index}]").get("optional_fields"),  # type: ignore[arg-type]
                extra_payload=_omit_keys(
                    _require_mapping(contract, f"command_contracts[{index}]"),
                    ("command", "required_fields", "optional_fields"),
                ),
            )
            for index, contract in enumerate(command_contracts)
        ],
    }
    if supported_entry_modes is not None:
        base["supported_entry_modes"] = _require_string_list(
            supported_entry_modes,
            "supported_entry_modes",
        )
    if _non_empty_text(product_entry_kind) is not None:
        base["product_entry_kind"] = _non_empty_text(product_entry_kind)
    return validate_family_domain_entry_contract(
        _merge_extra_payload(base, extra_payload, field="domain_entry_contract"),
        "domain_entry_contract",
    )


def validate_gateway_interaction_contract(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    return {
        **dict(payload),
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
        "frontdoor_owner": _require_string(payload.get("frontdoor_owner"), f"{field}.frontdoor_owner"),
        "user_interaction_mode": _require_string(
            payload.get("user_interaction_mode"),
            f"{field}.user_interaction_mode",
        ),
        "user_commands_required": _require_bool(
            payload.get("user_commands_required"),
            f"{field}.user_commands_required",
        ),
        "command_surfaces_for_agent_consumption_only": _require_bool(
            payload.get("command_surfaces_for_agent_consumption_only"),
            f"{field}.command_surfaces_for_agent_consumption_only",
        ),
        "shared_downstream_entry": _require_string(
            payload.get("shared_downstream_entry"),
            f"{field}.shared_downstream_entry",
        ),
        "shared_handoff_envelope": _require_string_list(
            payload.get("shared_handoff_envelope"),
            f"{field}.shared_handoff_envelope",
        ),
    }


def build_gateway_interaction_contract(
    *,
    frontdoor_owner: str,
    user_interaction_mode: str,
    user_commands_required: bool,
    command_surfaces_for_agent_consumption_only: bool,
    shared_downstream_entry: str,
    shared_handoff_envelope: list[str],
    surface_kind: str = "gateway_interaction_contract",
    extra_payload: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    return validate_gateway_interaction_contract(
        _merge_extra_payload(
            {
                "surface_kind": _require_string(surface_kind, "surface_kind"),
                "frontdoor_owner": _require_string(frontdoor_owner, "frontdoor_owner"),
                "user_interaction_mode": _require_string(
                    user_interaction_mode,
                    "user_interaction_mode",
                ),
                "user_commands_required": _require_bool(
                    user_commands_required,
                    "user_commands_required",
                ),
                "command_surfaces_for_agent_consumption_only": _require_bool(
                    command_surfaces_for_agent_consumption_only,
                    "command_surfaces_for_agent_consumption_only",
                ),
                "shared_downstream_entry": _require_string(
                    shared_downstream_entry,
                    "shared_downstream_entry",
                ),
                "shared_handoff_envelope": _require_string_list(
                    shared_handoff_envelope,
                    "shared_handoff_envelope",
                ),
            },
            extra_payload,
            field="gateway_interaction_contract",
        ),
        "gateway_interaction_contract",
    )


def validate_shared_handoff_builder(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    normalized = {
        **dict(payload),
        "command": _require_string(payload.get("command"), f"{field}.command"),
        "entry_mode": _require_string(payload.get("entry_mode"), f"{field}.entry_mode"),
    }
    surface_kind = _non_empty_text(payload.get("surface_kind"))
    if surface_kind is not None:
        normalized["surface_kind"] = surface_kind
    return normalized


def build_shared_handoff_builder(
    *,
    command: str,
    entry_mode: str,
    surface_kind: str | None = None,
    extra_payload: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    base: dict[str, Any] = {
        "command": _require_string(command, "command"),
        "entry_mode": _require_string(entry_mode, "entry_mode"),
    }
    normalized_surface_kind = _non_empty_text(surface_kind)
    if normalized_surface_kind is not None:
        base["surface_kind"] = normalized_surface_kind
    return validate_shared_handoff_builder(
        _merge_extra_payload(base, extra_payload, field="shared_handoff_builder"),
        "shared_handoff_builder",
    )


def validate_shared_handoff_return_surface(value: object, field: str) -> dict[str, Any]:
    payload = _require_mapping(value, field)
    return {
        **dict(payload),
        "surface_kind": _require_string(payload.get("surface_kind"), f"{field}.surface_kind"),
        "target_domain_id": _require_string(payload.get("target_domain_id"), f"{field}.target_domain_id"),
    }


def build_shared_handoff_return_surface(
    *,
    surface_kind: str,
    target_domain_id: str,
    extra_payload: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    return validate_shared_handoff_return_surface(
        _merge_extra_payload(
            {
                "surface_kind": _require_string(surface_kind, "surface_kind"),
                "target_domain_id": _require_string(target_domain_id, "target_domain_id"),
            },
            extra_payload,
            field="shared_handoff_return_surface",
        ),
        "shared_handoff_return_surface",
    )
